File: java/com/google/dotprompt/models/PromptMetadata.java
/*
 * Copyright 2025 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package com.google.dotprompt.models;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Metadata associated with a prompt definition.
 *
 * <p>This record contains all the configuration and metadata fields that can be specified in a
 * prompt's YAML frontmatter. It corresponds to the JS PromptMetadata interface.
 *
 * @param name The name of the prompt.
 * @param variant The variant name for the prompt.
 * @param version The version of the prompt.
 * @param description A description of the prompt.
 * @param model The name of the model to use (e.g., "vertexai/gemini-1.0-pro").
 * @param tools Names of tools (registered separately) to allow use of in this prompt.
 * @param toolDefs Definitions of tools to allow use of in this prompt.
 * @param config Model configuration. Not all models support all options.
 * @param input Configuration for input variables including defaults and schema.
 * @param output Defines the expected model output format and schema.
 * @param raw The raw frontmatter as parsed with no additional processing.
 * @param ext Extension fields gathered by namespace (fields containing a period).
 * @param metadata Arbitrary metadata for tooling or informational purposes.
 */
public record PromptMetadata(
    String name,
    String variant,
    String version,
    String description,
    String model,
    List<String> tools,
    List<ToolDefinition> toolDefs,
    Map<String, Object> config,
    InputConfig input,
    OutputConfig output,
    Map<String, Object> raw,
    Map<String, Map<String, Object>> ext,
    Map<String, Object> metadata) {

  /**
   * Configuration for input variables.
   *
   * @param defaultValues Default input variable values to use if none are provided.
   * @param schema Schema definition for input variables.
   */
  public record InputConfig(Map<String, Object> defaultValues, Object schema) {
    /** Alias for defaultValues to match JS naming. */
    public Map<String, Object> getDefault() {
      return defaultValues;
    }
  }

  /**
   * Configuration for output format.
   *
   * @param format Desired output format ("json", "text", or custom).
   * @param schema Schema defining the output structure.
   */
  public record OutputConfig(String format, Object schema) {}

  /** Creates empty metadata. */
  public PromptMetadata() {
    this(null, null, null, null, null, null, null, null, null, null, null, null, null);
  }

  /**
   * Creates metadata from a config map (typically from parsed frontmatter).
   *
   * @param config The configuration map.
   * @return A PromptMetadata instance.
   */
  @SuppressWarnings("unchecked")
  public static PromptMetadata fromConfig(Map<String, Object> config) {
    if (config == null || config.isEmpty()) {
      return new PromptMetadata();
    }

    String name = (String) config.get("name");
    String variant = (String) config.get("variant");
    String version = (String) config.get("version");
    String description = (String) config.get("description");
    String model = (String) config.get("model");
    List<String> tools = (List<String>) config.get("tools");
    List<ToolDefinition> toolDefs = null;
    Object toolDefsRaw = config.get("toolDefs");
    if (toolDefsRaw instanceof List) {
      toolDefs = (List<ToolDefinition>) toolDefsRaw;
    }
    Map<String, Object> modelConfig = (Map<String, Object>) config.get("config");

    InputConfig input = null;
    Object inputRaw = config.get("input");
    if (inputRaw instanceof Map) {
      @SuppressWarnings("unchecked")
      Map<String, Object> inputMap = (Map<String, Object>) inputRaw;
      // Schema can be a Map (JSON Schema) or other type (picoschema String)
      Object inputSchema = inputMap.get("schema");
      input = new InputConfig((Map<String, Object>) inputMap.get("default"), inputSchema);
    }

    OutputConfig output = null;
    Object outputRaw = config.get("output");
    if (outputRaw instanceof Map) {
      @SuppressWarnings("unchecked")
      Map<String, Object> outputMap = (Map<String, Object>) outputRaw;
      // Schema can be a Map (JSON Schema) or other type (picoschema String)
      Object outputSchema = outputMap.get("schema");
      output = new OutputConfig((String) outputMap.get("format"), outputSchema);
    }

    Map<String, Object> raw = (Map<String, Object>) config.get("raw");
    Map<String, Map<String, Object>> ext = (Map<String, Map<String, Object>>) config.get("ext");
    Map<String, Object> metadata = (Map<String, Object>) config.get("metadata");

    return new PromptMetadata(
        name,
        variant,
        version,
        description,
        model,
        tools,
        toolDefs,
        modelConfig,
        input,
        output,
        raw,
        ext,
        metadata);
  }

  /**
   * Converts this metadata to a config map (inverse of fromConfig).
   *
   * <p>This is useful when the metadata needs to be passed to methods that accept Map-based
   * configuration, such as the internal renderMetadata implementation.
   *
   * @return A map representation of this metadata.
   */
  public Map<String, Object> toConfig() {
    Map<String, Object> map = new HashMap<>();

    if (name != null) map.put("name", name);
    if (variant != null) map.put("variant", variant);
    if (version != null) map.put("version", version);
    if (description != null) map.put("description", description);
    if (model != null) map.put("model", model);
    if (tools != null) map.put("tools", tools);
    if (toolDefs != null) map.put("toolDefs", toolDefs);
    if (config != null) map.put("config", config);

    if (input != null) {
      Map<String, Object> inputMap = new HashMap<>();
      if (input.defaultValues() != null) inputMap.put("default", input.defaultValues());
      if (input.schema() != null) inputMap.put("schema", input.schema());
      map.put("input", inputMap);
    }

    if (output != null) {
      Map<String, Object> outputMap = new HashMap<>();
      if (output.format() != null) outputMap.put("format", output.format());
      if (output.schema() != null) outputMap.put("schema", output.schema());
      map.put("output", outputMap);
    }

    if (raw != null) map.put("raw", raw);
    if (ext != null) map.put("ext", ext);
    if (metadata != null) map.put("metadata", metadata);

    return map;
  }
}

File: dart/dotprompt/lib/src/models/prompt_metadata.dart
// Copyright 2026 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
//
// SPDX-License-Identifier: Apache-2.0

/// Prompt metadata representation.
///
/// Contains all configuration extracted from the YAML frontmatter of a
/// prompt file, including model settings, input/output schemas, and tools.
library;

import "package:meta/meta.dart";

import "parsed_prompt.dart";

/// Metadata parsed from prompt frontmatter.
///
/// This class contains all the configuration options that can be specified
/// in the YAML frontmatter of a .prompt file.
///
/// ## Frontmatter Fields
///
/// | Field    | Type                  | Description                        |
/// |----------|-----------------------|------------------------------------|
/// | model    | String                | Model identifier                   |
/// | config   | Map                   | Generation configuration           |
/// | input    | InputConfig           | Input schema and defaults          |
/// | output   | OutputConfig          | Output format and schema           |
/// | tools    | List\<String\>          | Tool names to enable               |
/// | ext.*    | Map                   | Extension fields                   |
///
/// ## Example Frontmatter
///
/// ```yaml
/// ---
/// model: gemini-pro
/// config:
///   temperature: 0.7
///   maxOutputTokens: 1024
/// input:
///   schema:
///     name: string
///     topic: string
///   default:
///     topic: "AI"
/// output:
///   format: json
///   schema:
///     summary: string
///     keyPoints: string[]
/// tools:
///   - searchWeb
///   - calculator
/// ext.custom:
///   myField: value
/// ---
/// ```
@immutable
class PromptMetadata {
  /// Creates a new [PromptMetadata].
  const PromptMetadata({
    this.model,
    this.config,
    this.input,
    this.output,
    this.tools,
    this.toolDefs,
    this.ext,
    this.raw,
  });

  /// Creates [PromptMetadata] from a configuration map.
  factory PromptMetadata.fromConfig(Map<String, dynamic> config) {
    final ext = <String, Map<String, dynamic>>{};

    // Extract extension fields (keys containing dots)
    for (final entry in config.entries) {
      if (entry.key.contains(".")) {
        final parts = entry.key.split(".");
        final extKey = parts[0];
        ext.putIfAbsent(extKey, () => {});
        if (parts.length == 2) {
          ext[extKey]![parts[1]] = entry.value;
        } else {
          // Handle nested ext keys like ext1.sub1.foo
          ext["${parts[0]}.${parts[1]}"] ??= {};
          ext["${parts[0]}.${parts[1]}"]![parts.sublist(2).join(".")] = entry.value;
        }
      }
    }

    return PromptMetadata(
      model: config["model"] as String?,
      config: config["config"] as Map<String, dynamic>?,
      input: config["input"] != null ? InputConfig.fromValue(config["input"]) : null,
      output: config["output"] != null ? OutputConfig.fromValue(config["output"]) : null,
      tools: config["tools"] != null ? (config["tools"] as List).cast<String>() : null,
      toolDefs: config["toolDefs"] != null
          ? (config["toolDefs"] as List).map((e) => ToolDefinition.fromJson(e as Map<String, dynamic>)).toList()
          : null,
      ext: ext.isNotEmpty ? ext : null,
      raw: config,
    );
  }

  /// Creates [PromptMetadata] from a JSON map.
  factory PromptMetadata.fromJson(Map<String, dynamic> json) => PromptMetadata.fromConfig(json);

  /// The model identifier.
  final String? model;

  /// Generation configuration (temperature, maxTokens, etc.).
  final Map<String, dynamic>? config;

  /// Input configuration.
  final InputConfig? input;

  /// Output configuration.
  final OutputConfig? output;

  /// List of tool names to enable.
  final List<String>? tools;

  /// Resolved tool definitions.
  final List<ToolDefinition>? toolDefs;

  /// Extension fields.
  final Map<String, Map<String, dynamic>>? ext;

  /// Raw frontmatter as parsed.
  final Map<String, dynamic>? raw;

  /// Converts this to a configuration map.
  Map<String, dynamic> toConfig() {
    final result = <String, dynamic>{};
    if (model != null) result["model"] = model;
    // Spread config values at top level (temperature, maxTokens, etc.)
    if (config != null) result.addAll(config!);
    if (raw != null) result["raw"] = raw;
    if (input != null) result["input"] = input!.toJson();
    if (output != null) result["output"] = output!.toJson();
    if (tools != null) result["tools"] = tools;
    if (toolDefs != null) {
      result["toolDefs"] = toolDefs!.map((t) => t.toJson()).toList();
    }
    if (ext != null) {
      result["ext"] = ext;
    }
    return result;
  }

  /// Converts this to a JSON-serializable map.
  Map<String, dynamic> toJson() => toConfig();

  @override
  bool operator ==(Object other) =>
      identical(this, other) || other is PromptMetadata && model == other.model && tools == other.tools;

  @override
  int get hashCode => Object.hash(model, tools);

  @override
  String toString() => "PromptMetadata(model: $model, config: $config, "
      "input: $input, output: $output, tools: $tools)";
}

/// Input configuration for a prompt.
@immutable
class InputConfig {
  /// Creates a new [InputConfig].
  const InputConfig({this.schema, this.defaultValues});

  /// Creates an [InputConfig] from a JSON map.
  factory InputConfig.fromJson(Map<String, dynamic> json) {
    final schemaValue = json["schema"];
    Map<String, dynamic>? schema;
    if (schemaValue is String) {
      // Schema is a type string (Picoschema), use as-is for later conversion
      schema = {r"$type": schemaValue};
    } else if (schemaValue is Map) {
      schema = schemaValue.cast<String, dynamic>();
    }
    return InputConfig(
      schema: schema,
      defaultValues: json["default"] as Map<String, dynamic>?,
    );
  }

  /// Creates an [InputConfig] from a value that can be a String or Map.
  ///
  /// If the value is a String, it's treated as a schema name reference.
  factory InputConfig.fromValue(dynamic value) {
    if (value is String) {
      // String value is a schema name reference
      return InputConfig(schema: {r"$ref": value});
    } else if (value is Map<String, dynamic>) {
      return InputConfig.fromJson(value);
    }
    throw ArgumentError.value(value, "value", "Expected String or Map");
  }

  /// JSON Schema for input validation.
  final Map<String, dynamic>? schema;

  /// Default values for input variables.
  final Map<String, dynamic>? defaultValues;

  /// Converts this to a JSON-serializable map.
  Map<String, dynamic> toJson() => {
        if (schema != null) "schema": schema,
        if (defaultValues != null) "default": defaultValues,
      };

  @override
  bool operator ==(Object other) => identical(this, other) || other is InputConfig;

  @override
  int get hashCode => Object.hash(schema, defaultValues);

  @override
  String toString() => "InputConfig(schema: $schema, defaultValues: $defaultValues)";
}

/// Output configuration for a prompt.
@immutable
class OutputConfig {
  /// Creates a new [OutputConfig].
  const OutputConfig({this.format, this.schema});

  /// Creates an [OutputConfig] from a JSON map.
  factory OutputConfig.fromJson(Map<String, dynamic> json) {
    final schemaValue = json["schema"];
    Map<String, dynamic>? schema;
    if (schemaValue is String) {
      // Schema is a type string (Picoschema), use as-is for later conversion
      schema = {r"$type": schemaValue};
    } else if (schemaValue is Map) {
      schema = schemaValue.cast<String, dynamic>();
    }
    return OutputConfig(format: json["format"] as String?, schema: schema);
  }

  /// Creates an [OutputConfig] from a value that can be a String or Map.
  ///
  /// If the value is a String, it's treated as a schema name reference.
  factory OutputConfig.fromValue(dynamic value) {
    if (value is String) {
      // String value is a schema name reference
      return OutputConfig(schema: {r"$ref": value});
    } else if (value is Map<String, dynamic>) {
      return OutputConfig.fromJson(value);
    }
    throw ArgumentError.value(value, "value", "Expected String or Map");
  }

  /// Output format (e.g., "json", "text").
  final String? format;

  /// JSON Schema for output validation.
  final Map<String, dynamic>? schema;

  /// Converts this to a JSON-serializable map.
  Map<String, dynamic> toJson() => {
        if (format != null) "format": format,
        if (schema != null) "schema": schema,
      };

  @override
  bool operator ==(Object other) => identical(this, other) || other is OutputConfig;

  @override
  int get hashCode => Object.hash(format, schema);

  @override
  String toString() => "OutputConfig(format: $format, schema: $schema)";
}
